src/matrix_ops.cpp

```diff
@@ -0,0 +1,189 @@
+/*
+ *
+ * matrix_ops.cpp
+ * Distance matrix transformations
+ *
+ */
+#include <vector>
+#include <thread>
+#include <cstdint>
+#include <cstddef>
+#include <cmath>
+#include <string>
+
+#include "api.hpp"
+
+// Prototypes for cppying functions run in threads
+void square_block(const Eigen::VectorXf& longDists,
+                  SquareMatrix& squareMatrix,
+                  const size_t n_samples,
+                  const size_t start,
+                  const size_t offset,
+                  const size_t max_elems);
+
+void rectangle_block(const Eigen::VectorXf& longDists,
+                  SquareMatrix& squareMatrix,
+                  const size_t nrrSamples,
+                  const size_t nqqSamples,
+                  const size_t start,
+                  const size_t max_elems);
+
+template<class T>
+size_t rows_to_samples(const T& longMat) {
+    return 0.5*(1 + sqrt(1 + 8*(longMat.rows())));
+}
+
+long calc_row_idx(const long long k, const long n) {
+	return n - 2 - floor(sqrt((double)(-8*k + 4*n*(n-1)-7))/2 - 0.5);
+}
+
+long calc_col_idx(const long long k, const long i, const long n) {
+	return k + i + 1 - n*(n-1)/2 + (n-i)*((n-i)-1)/2;
+}
+
+long long square_to_condensed(long i, long j, long n) {
+    assert(i > j);
+	return (n*j - ((j*(j+1)) >> 1) + i - 1 - j);
+}
+
+std::tuple<size_t, unsigned int, unsigned int> 
+   jobs_per_thread(const size_t num_jobs, 
+                   const unsigned int num_threads) {
+    size_t used_threads = num_threads;
+    if (num_jobs < used_threads) {
+        used_threads = num_jobs; 
+    }
+    size_t calc_per_thread = (size_t)num_jobs / used_threads;
+    unsigned int num_big_threads = num_jobs % used_threads;
+
+    return(std::make_tuple(calc_per_thread, used_threads, num_big_threads)); 
+}
+
+SquareMatrix long_to_square(const Eigen::VectorXf& rrDists, 
+                            const Eigen::VectorXf& qrDists,
+                            const Eigen::VectorXf& qqDists,
+                            unsigned int num_threads) {
+    // Set up square matrix to move values into
+    size_t nrrSamples = rows_to_samples(rrDists);
+    size_t nqqSamples = 0;
+    if (qrDists.size() > 0) {
+        nqqSamples = rows_to_samples(qqDists);
+        if (qrDists.size() != nrrSamples * nqqSamples) {
+            throw std::runtime_error("Shape of reference, query and ref vs query matrices mismatches");
+        }
+    }
+    SquareMatrix squareDists = SquareMatrix::Zero(nrrSamples + nqqSamples, nrrSamples + nqqSamples);
```

Comment:
TODO: remove this and set diagonal to zero at the end (might be a bit wasteful setting every element twice)